### PR TITLE
[ADD] document_ocr

### PR DESCRIPTION
**Pull request by hbrunn**
# 
# OCR for documents

This module was written to make uploaded documents, for example scans, searchable by running OCR on them.

It supports all image formats `Pillow supports <http://pillow.readthedocs.io/en/3.2.x/handbook/image-file-formats.html>`_ for reading and PDFs.
# Installation

To install this module, you need to:
# . install tesseract and the language(s) your documents use
# . if you want to support OCR on PDFs, install imagemagick
# . install the module itself
# Configuration

To configure this module, go to:
# . Settings/Technical/Parameters/System parameters and review the parameters with names document_ocr.*
# Usage

By default, character recognition is done asynchronously by a cronjob at night. This is because the recognition process takes a while and you don't want to make your users wait for the indexation to finish. If you want to force that nonetheless, set configuration parameter `document_ocr.synchronous` to value `True`.

### CODE REVIEW

File: document_ocr/models/ir_attachment.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                data, datas_fname, file_type, dpi)
+        else:
+            image_data = StringIO()
+            try:
+                Image.open(StringIO(data)).save(image_data, 'tiff',
+                                                dpi=(dpi, dpi))
+            except IOError:
+                _logger.exception('Failed to OCR image')
+                return None
+        process = subprocess.Popen(
```
Comment (vperilla) [inline document_ocr/models/ir_attachment.py:50]: Why not use pytesseract?

Comment (hbrunn) [inline document_ocr/models/ir_attachment.py:50]: what benefit would we get?

File: document_ocr/README.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+It supports all image formats `Pillow supports <http://pillow.readthedocs.io/en/3.2.x/handbook/image-file-formats.html>`_ for reading and PDFs.
+
+Installation
+============
+
+To install this module, you need to:
+
+#. install tesseract and the language(s) your documents use
+#. if you want to support OCR on PDFs, install imagemagick
+#. install the module itself
```
Comment (dreispt) [inline document_ocr/README.rst:20]: It would be helpful to put the actual debian apt-get command here.
What about the `fonts-inconsolata` dependency?

File: document_ocr/README.rst [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+=============
+
+To configure this module, go to:
+
+#. Settings/Technical/Parameters/System parameters and review the parameters with names document_ocr.*
+
+Usage
+=====
+
+By default, character recognition is done asynchronously by a cronjob at night. This is because the recognition process takes a while and you don't want to make your users wait for the indexation to finish. If you want to force that nonetheless, set configuration parameter ``document_ocr.synchronous`` to value ``True``.
```
Comment (dreispt) [inline document_ocr/README.rst]: > By default, character recognition is done asynchronously by a cronjob at night.

Can we tune this parameter,for example, to make it run every x hours?

File: document_ocr/models/ir_attachment.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+# © 2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import logging
+import subprocess
+from PIL import Image
+from StringIO import StringIO
+from openerp import api, models
+
+_logger = logging.getLogger(__name__)
+_marker_phrase = '[[waiting for OCR]]'
```
Comment (dreispt) [inline document_ocr/models/ir_attachment.py]: Should this go in uppercase?

File: document_ocr/README.rst (hunk 3 above)
Comment (hbrunn) [inline document_ocr/README.rst]: sure, it's an `ir.cron`

File: document_ocr/README.rst (hunk 2 above)
Comment (hbrunn) [inline document_ocr/README.rst:20]: good idea, will do. The font is only needed for the test, because there I generate a picture with image magick. And the default font (what comes if there are no fonts on the system) is so small and pixely that tesseract can't recognize it

**Comment by dreispt** [comment]
See https://github.com/hbrunn/knowledge/pull/1

File: document_ocr/models/ir_attachment.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate(data)
+        if stderr:
+            _logger.error('Error converting to PDF: %s', stderr)
+        return StringIO(stdout)
+
+    @api.model
+    def _ocr_cron(self):
+        for this in self.with_context(document_ocr_force=True).search([
```
Comment (dreispt) [inline document_ocr/models/ir_attachment.py:74]: Alert! Javascripter spotted! 😄

File: .travis.yml [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
 cache: pip
 
 addons:
   apt:
     packages:
       - expect-dev  # provides unbuffer utility
       - python-lxml # because pip installation is slow
+      - tesseract-ocr # document_ocr
+      - imagemagick # document_ocr
+      - fonts-inconsolata # document_ocr
```
Comment (dreispt) [inline .travis.yml]: This is only to tests, right? If so, it's worth mentioning it in the comment

File: document_ocr/README.rst (hunk 3 above)
Comment (dreispt) [inline document_ocr/README.rst]: Could you add a short sentence mentioning that, please?

**Comment by dreispt** [comment]
:+1:

File: document_ocr/models/ir_attachment.py (hunk 5 above)
Comment (hbrunn) [inline document_ocr/models/ir_attachment.py:74]: actually, I was socialized with C, where I formatted stuff like this from the beginning. But indeed, the habit of doing the same in Python is new and derived from doing a lot of JS work lately.

I won't be able to pick up your comments soon, so if you're willing to do another PR you'd speedup things a lot.

Comment (dreispt) [inline document_ocr/models/ir_attachment.py:74]: OK. As you may have noticed, I'm experimenting on a new way to do reviews.
My PR here was done exclusively through the GitHub web interface.

Comment (hbrunn) [inline document_ocr/models/ir_attachment.py:74]: seems a good way to me. Then my excuse doesn't work, because I don't have access to my github SSH key at the moment :-(
but if you want to practice a second time with the github interface...

**Comment by dreispt** [comment]
See https://github.com/hbrunn/knowledge/pull/2

**Comment by dreispt** [comment]
Just noticed that tests are [not providing effective coverage](https://travis-ci.org/OCA/knowledge/jobs/142774820#L724):
the lines missing are the ones implementing the actual OCR logic.

**Comment by hbrunn** [comment]
that's actually caused by me not reading your PR closely enough. You missed uppercasing the marker phrase in the test: https://travis-ci.org/OCA/knowledge/jobs/142774820#L603

**Comment by dreispt** [comment]
Damn reviewers!
See https://github.com/hbrunn/knowledge/pull/3

**Comment by hbrunn** [comment]
damn reviewers, indeed ;-)

**Comment by hbrunn** [comment]
@dreispt one occurrence left: https://travis-ci.org/OCA/knowledge/jobs/142788839#L335

**Comment by dreispt** [comment]
https://github.com/hbrunn/knowledge/pull/4

**Comment by hbrunn** [comment]
yay, it's green again, thanks!

File: document_ocr/models/ir_attachment.py (hunk 1 above)
Comment (lasley) [inline document_ocr/models/ir_attachment.py:50]: Mainly just line reduction and more Pythonic usage. It's just a shell wrapper though, so beyond developer usability & error handling - technically nothing.

```
from tesseract import image_to_string
image_to_string(Image.open(), lang='eng')
```

Comment (hbrunn) [inline document_ocr/models/ir_attachment.py:50]: that's what my question was meant to imply: We gain basically nothing, and that's not worth an extra dependency. If this were python bindings to libtesseract, it was a completely different story.

**Comment by dreispt** [comment]
Good. Module [coverage](https://travis-ci.org/OCA/knowledge/jobs/142806516#L716) is now 93%.

**Comment by lasley** [comment]
Thanks for the awesome module @hbrunn (and subsequent fixes @dreispt ) - interesting that you seem to be just a few steps ahead of all of my projects 😛 

Works in my local + code review 👍

**Comment by hbrunn** [comment]
@gurneyalex is it possible to install `tesseract-ocr` on runbot? Without this, this repo would become permanently red on runbot after merging this, and that would be a shame of course.

**Comment by dreispt** [comment]
@lasley A possible strategy is to open a WIP PR starting with just the README. That advertises that you are working on something, and you can get an early discussion on your design.

File: document_ocr/README.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+    :alt: Try me on Runbot
+    :target: https://runbot.odoo-community.org/runbot/118/8.0
+
+For further information, please visit:
```
Comment (elicoidal) [inline document_ocr/README.rst]: You can remove this part (newest README template)

File: document_ocr/README.rst [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+* https://www.odoo.com/forum/help-1
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/knowledge/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/knowledge/issues/new?body=module:%20document_ocr%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (elicoidal) [inline document_ocr/README.rst]: You can remove this part (newest README template)

File: document_ocr/README.rst [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* Holger Brunn <hbrunn@therp.nl>  
+
+Do not contact contributors directly about help with questions or problems concerning this addon, but use the `community mailing list <mailto:community@mail.odoo.com>`_ or the `appropriate specialized mailinglist <https://odoo-community.org/groups>`_ for help, and the bug tracker linked in `Bug Tracker`_ above for technical issues.
```
Comment (elicoidal) [inline document_ocr/README.rst:71]: You can remove this part (check newest README template)

File: document_ocr/__openerp__.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    "summary": "Run character recognition on uploaded files",
+    "depends": [
+        'document',
+    ],
+    "data": [
+        "data/ir_cron.xml",
+        "data/ir_config_parameter.xml",
+    ],
+    "external_dependencies": {
+        'python': [],
```
Comment (elicoidal) [inline document_ocr/__openerp__.py]: Please remove empty keys

File: document_ocr/models/ir_attachment.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+# © 2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import logging
+import subprocess
+from PIL import Image
```
Comment (elicoidal) [inline document_ocr/models/ir_attachment.py:6]: Should not you protect the import(s)?

**Comment by dreispt** [comment]
again :+1:

File: document_ocr/models/ir_attachment.py (hunk 11 above)
Comment (hbrunn) [inline document_ocr/models/ir_attachment.py:6]: this one is a dependency of odoo itself, so no

**Comment by gurneyalex** [comment]
@hbrunn tesseract-ocr installed, and rebuild forced.

File: .travis.yml [hunk 12]
```diff
@@ -1,11 +1,15 @@
-sudo: false
+sudo: required
```
Comment (gurneyalex) [inline .travis.yml:1]: I'm missing the need for the above change. Can you enlighten me?

**Comment by gurneyalex** [comment]
globally :+1: 

Thanks for this great idea :smile_cat:

**Comment by dreispt** [comment]
Need comments from @elicoidal

File: .travis.yml (hunk 12 above)
Comment (hbrunn) [inline .travis.yml:1]: this was some problem with using trusty for tests. Just pushed a new commit reverting this to see what happens

Comment (hbrunn) [inline .travis.yml:1]: @gurneyalex that's why then: https://travis-ci.org/OCA/knowledge/jobs/148364064 - the docker image still has an older version of tesseract that doesn't work well with the module

**Comment by hbrunn** [comment]
done and rebased

**Comment by rafaelbn** [comment]
Super! :+1:

**Comment by lasley** [comment]
Code + functional 👍 

image

**Comment by pedrobaeza** [comment]
No more reason to retain this as PR. Thanks for another superb module!
